embedded/store/indexer.go

```diff
@@ -371,78 +420,100 @@ func (idx *indexer) doIndexing() {
-	err := idx.store.readTx(txID, false, idx.tx)
-	if err != nil {
-		return err
-	}
-
-	txEntries := idx.tx.Entries()
+func (idx *indexer) indexSince(txID uint64) error {
+	ctx, cancel := context.WithTimeout(context.Background(), idx.bulkPreparationTimeout)
+	defer cancel()
 
-	var txmd []byte
+	bulkSize := 0
+	indexableEntries := 0
 
-	if idx.tx.header.Metadata != nil {
-		txmd = idx.tx.header.Metadata.Bytes()
-	}
+	for i := 0; i < idx.maxBulkSize; i++ {
+		err := idx.store.readTx(txID+uint64(i), false, idx.tx)
+		if err != nil {
+			return err
+		}
 
-	txmdLen := len(txmd)
+		txEntries := idx.tx.Entries()
 
-	indexableEntries := 0
+		var txmd []byte
 
-	for _, e := range txEntries {
-		if e.md != nil && e.md.NonIndexable() {
-			continue
+		if idx.tx.header.Metadata != nil {
+			txmd = idx.tx.header.Metadata.Bytes()
 		}
 
-		// vLen + vOff + vHash + txmdLen + txmd + kvmdLen + kvmd
-		var b [lszSize + offsetSize + sha256.Size + sszSize + maxTxMetadataLen + sszSize + maxKVMetadataLen]byte
-		o := 0
+		txmdLen := len(txmd)
+
+		for _, e := range txEntries {
+			if e.md != nil && e.md.NonIndexable() {
+				continue
+			}
 
-		binary.BigEndian.PutUint32(b[o:], uint32(e.vLen))
-		o += lszSize
+			// vLen + vOff + vHash + txmdLen + txmd + kvmdLen + kvmd
+			var b [lszSize + offsetSize + sha256.Size + sszSize + maxTxMetadataLen + sszSize + maxKVMetadataLen]byte
+			o := 0
 
-		binary.BigEndian.PutUint64(b[o:], uint64(e.vOff))
-		o += offsetSize
+			binary.BigEndian.PutUint32(b[o:], uint32(e.vLen))
+			o += lszSize
 
-		copy(b[o:], e.hVal[:])
-		o += sha256.Size
+			binary.BigEndian.PutUint64(b[o:], uint64(e.vOff))
+			o += offsetSize
 
-		binary.BigEndian.PutUint16(b[o:], uint16(txmdLen))
-		o += sszSize
+			copy(b[o:], e.hVal[:])
+			o += sha256.Size
 
-		copy(b[o:], txmd)
-		o += txmdLen
+			binary.BigEndian.PutUint16(b[o:], uint16(txmdLen))
+			o += sszSize
 
-		var kvmd []byte
+			copy(b[o:], txmd)
+			o += txmdLen
 
-		if e.md != nil {
-			kvmd = e.md.Bytes()
-		}
+			var kvmd []byte
+
+			if e.md != nil {
+				kvmd = e.md.Bytes()
+			}
 
-		kvmdLen := len(kvmd)
+			kvmdLen := len(kvmd)
 
-		binary.BigEndian.PutUint16(b[o:], uint16(kvmdLen))
-		o += sszSize
+			binary.BigEndian.PutUint16(b[o:], uint16(kvmdLen))
+			o += sszSize
 
-		copy(b[o:], kvmd)
-		o += kvmdLen
+			copy(b[o:], kvmd)
+			o += kvmdLen
 
-		idx.store._kvs[indexableEntries].K = e.key()
-		idx.store._kvs[indexableEntries].V = b[:o]
+			idx._kvs[indexableEntries].K = e.key()
+			idx._kvs[indexableEntries].V = b[:o]
+			idx._kvs[indexableEntries].T = txID + uint64(i)
 
-		indexableEntries++
+			indexableEntries++
+		}
+
+		bulkSize++
+
+		if bulkSize < idx.maxBulkSize {
+			// wait for the next tx to be committed
+			err = idx.store.commitWHub.WaitFor(txID+uint64(i+1), ctx.Done())
+		}
+		if err == watchers.ErrCancellationRequested {
+			break
+		}
+		if err != nil {
+			return err
+		}
 	}
 
+	var err error
+
 	if indexableEntries == 0 {
```

Comment:
Can you just add a comment on why ts is being increased when indexableEntries ==0? :)

Reply:
when a transaction didn't have any indexable entry, the logical time in the tree was increased. But that is actually not required anymore, given the transaction id is specified per entry now